consensus/tendermint/core/message.go

```diff
@@ -123,15 +139,25 @@ func (m *Message) FromPayload(b []byte, previousHeader *types.Header, validateFn
 
 	v := previousHeader.CommitteeMember(addr)
 	if v == nil {
-		return nil, fmt.Errorf("validator was not a committee member %q", v)
+		return nil, fmt.Errorf("message received is not from a committee member: %x", addr)
 	}
 
 	m.power = v.VotingPower.Uint64()
 	return v, nil
 }
 
-func (m *Message) Payload() ([]byte, error) {
-	return rlp.EncodeToBytes(m)
+func (m *Message) Payload() []byte {
+	if m.payload == nil {
+		payload, err := rlp.EncodeToBytes(m)
+		if err != nil {
+			// We panic if there is an error, reasons:
+			// Either we received the message and we managed to decode it, hence it must be possible to encode it.
+			// If we can't encode the payload for our own generated messages, that's a programming error.
+			panic("could not decode message payload")
+		}
+		m.payload = payload
+	}
+	return m.payload
```

Comment:
This is not safe for concurrent use.

Instead of checking for nil we can use `sync.Once`

```suggestion
	m.once.Do(func() {
		payload, err := rlp.EncodeToBytes(m)
		if err != nil {
			// We panic if there is an error, reasons:
			// Either we received the message and we managed to decode it, hence it must be possible to encode it.
			// If we can't encode the payload for our own generated messages, that's a programming error.
			panic("could not decode message payload")
		}
	})
	m.payload = payload
```

Reply:
yes it would appear that the `Payload()` method is not thread safe, but as far as I can see it is never called concurrently. The messages used to synchronize peers in `SyncPeers` have always this payload field being set by the main thread, before being saved in the current height message buffer.
I would like to avoid using unnecessary mutexes and I have run tests with the `-race` flag, couldn't see anything

Reply:
Its called in SyncPeer and in mainEventLoop. Those are 2 separate goroutines.

Reply:
What I'm trying to say it that when it SyncPeer is called m.payload is always set but let's discuss at standup.

Reply:
So that may be the case right now, but there is nothing to stop someone from changing the code and making it not so. It's hard to verify that this code is not called concurrently and there  is no hint in the code that we need to be careful about that.

The neatest way to handle this in my opinion is to avoid mutable messages. That means providing constructor functions that make sure all fields are set, which would ensure that you cannot have an instance of Message without all its fields being defined.